### PR TITLE
Speed hacks

### PR DESCRIPTION
**Pull request by ZacharyTaylor**
Looked at getting the cpu usage running voxblox down and ended up doing some more optimization.

There are five changes in this pr to speed things up. Let me know if you disagree with these changes (especially 1 and 2 as they could be considered over-optimization)

1) Switched from doubles to floats as the FloatingPoint type. This required changing types in CameraModel

2) When running the fast integrator getLocalFromGlobalVoxelIndex is the third most expensive function in voxblox (1. ApproxHashSet::replaceHash, 2. TsdfIntegratorBase::updateTsdfVoxel). This is due to the modulus operator. I constrained voxels_per_side to be a power of two and replaced this with a bitwise and.

3) Removed the pointer used to index the approx_array and just added an offset. Code is a bit cleaner and a bit faster

4) We were filtering nans twice in pointcloud preprocessing, removed the unneeded one.

5) Add flag to let the fast integrator use the same collision checking for multiple frames. Will give much nosier results but speeds things up if needed (just set to clear checks every frame by default)

Timings on one of the ros bags from jays flight last week:
Before:
inserting_missed_blocks     149  00.014100    
integrate                              149  06.616214   
ptcloud_preprocess             149  00.741457       

After:
inserting_missed_blocks     149  00.012957      
integrate                              149  05.251077       
ptcloud_preprocess             149  00.726723

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -178,14 +178,17 @@ A summary of the user setable voxblox_node parameters:
...
-| `tsdf_voxels_per_side` | TSDF voxels per side of an allocated block | 16 |
+| `tsdf_voxels_per_side` | TSDF voxels per side of an allocated block. Must be a power of 2 | 16 |
 | `voxel_carving_enabled` | If true, the entire length of a ray is integrated, if false only the region inside the trunaction distance is used. | true |
 | `truncation_distance` | The truncation distance for the TSDF | 2`tsdf_voxel_size` |
 | `max_ray_length_m` | The maximum range out to which a ray will be cast | 5.0 |
 | `min_ray_length_m` | The point at which the ray casting will start | 0.1 |
 | `max_weight` | The upper limit for the weight assigned to a voxel | 10000.0 |
 | `use_const_weight` | If true all points along a ray have equal weighting | false |
 | `allow_clear` | If true points beyond the `max_ray_length_m` will be integrated up to this distance | true |
+| `start_voxel_subsampling_factor` | Only used by the fast integrator. Before integration points are inserted into a sub-voxel, only one point is allowed per sub-voxel. This can be thought of as subsampling the pointcloud. The edge length of the sub-voxel is the voxel edge length divided by `start_voxel_subsampling_factor`. | 2 |
```
Comment (helenol) [inline README.md]: Nice!!! thanks for adding these.

File: voxblox/src/utils/camera_model.cc [hunk 2]
```diff
@@ -55,31 +53,31 @@ void CameraModel::setIntrinsicsFromFoV(double horizontal_fov,
...
+            -min_distance * tan_half_vertical_fov));
   untransformed_corners_.emplace_back(
-      Eigen::Vector3d(min_distance, -min_distance * tan_half_horizontal_fov,
-                      -min_distance * tan_half_vertical_fov));
+      Point(min_distance, -min_distance * tan_half_horizontal_fov,
+            -min_distance * tan_half_vertical_fov));
   untransformed_corners_.emplace_back(
-      Eigen::Vector3d(min_distance, -min_distance * tan_half_horizontal_fov,
-                      min_distance * tan_half_vertical_fov));
+      Point(min_distance, -min_distance * tan_half_horizontal_fov,
```
Comment (helenol) [inline voxblox/src/utils/camera_model.cc]: Ugh thanks for cleaning all these up.

File: voxblox_ros/src/tsdf_server.cc [hunk 3]
```diff
@@ -183,15 +191,17 @@ void TsdfServer::insertPointcloud(
 
     timing::Timer ptcloud_timer("ptcloud_preprocess");
 
-    // Filter out NaNs. :|
-    std::vector<int> indices;
-    pcl::removeNaNFromPointCloud(pointcloud_pcl, pointcloud_pcl, indices);
```
Comment (helenol) [inline voxblox_ros/src/tsdf_server.cc]: Also nice to swap this out; this function ironically only gets some nans.

File: voxblox/src/utils/camera_model.cc (hunk 2 above)
Comment (ZacharyTaylor) [reply voxblox/src/utils/camera_model.cc]: Was actually pretty quick just did a find replace for Eigen::Vector3d and Point

**Comment by mfehr** [comment]
@ZacharyTaylor thanks for yet another speed optimization! However optimization 1 just broke all the interfaces with whoever linked against voxblox... (which we do in maplab)

**Pull request merged**